Libraries/LibWeb/CSS/SelectorEngine.cpp

```diff
@@ -755,6 +755,9 @@ static ALWAYS_INLINE bool matches_namespace(
 
 static inline bool matches(CSS::Selector::SimpleSelector const& component, Optional<CSS::CSSStyleSheet const&> style_sheet_for_rule, DOM::Element const& element, GC::Ptr<DOM::Element const> shadow_host, GC::Ptr<DOM::ParentNode const> scope, SelectorKind selector_kind, [[maybe_unused]] GC::Ptr<DOM::Element const> anchor)
 {
+    if (shadow_host && &element == shadow_host.ptr() && component.type != CSS::Selector::SimpleSelector::Type::PseudoClass && component.type != CSS::Selector::SimpleSelector::Type::PseudoElement) {
+        return false;
+    }
```

Comment:
It'd be nice to move this check into its own static function, to save having to remember to modify both places.

Also... I think this doesn't block enough. It'll allow other kinds of pseudo-class, such as `:first-child`.

Reply:
updated it to allow only host and compound pseudo classes